### ISSUE TITLE
Detecting when connections are closed

### ISSUE THREAD

**Issue by neongreen** [snoyberg/http-client#394]
tl;dr My goal is to get some action to reliably execute whenever an http-client connection times out. I tried modifying `connectionClose`, but without luck.

---

Example here: <https://github.com/neongreen/http-client-destroy>.

* Start listening on port :12777 (`nc -l -k 12777`).
* Run my example and press Enter several times.
* A POST request is made on each Enter with a timeout of 5 seconds.
* When the connection is closed (`connectionClose`), "destroying" is printed.

I expected that every connection that is timing out would be followed by "destroying". In reality, it's very inconsistent. Sometimes I get "destroying" for a connection, especially if it's the first or second connection, but mostly I don't.

The meat of the example is as follows:

```haskell
    manager <- newManager defaultManagerSettings
        { managerResponseTimeout = responseTimeoutMicro 5000000
        , managerRawConnection = do
              mkConn <- managerRawConnection defaultManagerSettings
              pure $ \ha s i -> onDestroy (putStrLn "destroying") <$> mkConn ha s i
        }
    runResourceT $ for_ [(1::Int) ..] $ \i -> do
        _ <- liftIO getLine
        initialRequest <- parseRequest "POST http://localhost:12777"
        let request = initialRequest
                { requestBody = RequestBodyLBS (BSC.pack (show i))
                }
        forkIO $ do
            response <- liftResourceT (http request manager)
            liftResourceT (responseBody response `connect` sinkNull)
```

What am I doing wrong? Or is it a bug in http-client?

**Comment by snoyberg** [comment]
I'm not able to look into the details of this right now, sorry. If you're trying to determine if there's a bug in http-client, I'd be interested in knowing if there's a mismatched number of calls to connection open and close functions.

**Comment by neongreen** [comment]
> if there's a mismatched number of calls to connection open and close functions

It would seem so, yes.

**Comment by Yuras** [comment]
@neongreen Note that `http-client` relies on week finalizers to deallocate connections, so you shouldn't expect immediate deallocation. Try to insert `performMajorGC` after each `getLine`, and you'll see more consistent behavior.

We can make resource deallocation more reliable (probably not 100% reliable though), but it will be a feature request then, and you'll need to convince @snoyberg

**Comment by neongreen** [comment]
> Note that http-client relies on week finalizers to deallocate connections, so you shouldn't expect immediate deallocation.

Oh. That makes sense, yeah.

**Comment by snoyberg** [comment]
I totally missed that point, thanks @Yuras.

@neongreen Should we close this issue?
